framework/src/play-jdbc-evolutions/src/main/scala/play/api/db/evolutions/ApplicationEvolutions.scala

```
@@ -85,17 +85,20 @@ class ApplicationEvolutions @Inject() (
 
   private def withLock(db: Database, dbConfig: EvolutionsDatasourceConfig)(block: => Unit): Unit = {
     if (dbConfig.useLocks) {
-      val ds = db.dataSource
-      val url = db.url
-      val c = ds.getConnection
-      c.setAutoCommit(false)
-      val s = c.createStatement()
-      createLockTableIfNecessary(url, c, s, dbConfig)
-      lock(url, c, s, dbConfig)
-      try {
-        block
-      } finally {
-        unlock(c, s)
+      db.withTransaction { connection =>
+        val statement = connection.createStatement()
+        try {
+          val url = db.url
+          createLockTableIfNecessary(url, connection, statement, dbConfig)
+          lock(url, connection, statement, dbConfig)
+        } finally {
+          statement.close()
+        }
+        try {
+          block
+        } finally {
+          connection.commit()
```

Just noticed this.

`withTransaction` will do the commit/rollback to you, so this is not necessary and could result in unexpected behaviors (calling commit after rollback, for example).

---

Calling commit after rollback is a no-op, and will not cause unexpected behavior. It's written this way so that failures in `block` still result in a commit (to preserve the original behavior of the code) but other failures will result in a rollback (because I can't imagine anything else being sane).

Frankly, this evolutions code is a mess. Why doesn't `block` accept a connection, so its changes can be part of the same transaction? Why did the `unlock` method silently swallow errors? I'm limiting the scope of my changes to fix the obvious bugs while keeping as much of the original behavior as possible, but I have no faith that the original behavior makes sense. Some one needs to go through this code, understand it, and rewrite it with error handling in mind. But that is out of scope for fixing the original error.

---

Even if this useless commit was there before, would remove it in this refactoring (Boy scout code).

---

@cchantep What makes you think the `commit` call is useless?

---

Any `commit` in `withTransaction` will ne useless as the goal is to manage it automatically (see comments).

---

@cchantep What do you believe will happen if the `block` call throws an exception? How does the behavior differ with and without the extra try/catch/commit wrapper?

---

It is the responsibility of `withTransaction` to manage the transaction (rollback/commit) according the application of the block on the `connection`.

If that is not the case, `withTransaction` is wrong, if that is OK commit/rollback in the block are useless (as Also pointed out by Marcos).

---

`withTransaction` does do some management of the transaction, but it is limited to exception handling that escapes the block passed to it. It's intention is not to do all transaction management. There are things it can't do, like needing to rollback the transaction based on data read or interaction with a 2nd database or another service.

Of course, this is just exception handling, so `withTransaction` can handle it _if we are willing to change the behavior_. Neither you nor @marcospereira have been able to give a "Yes, the original code was wrong, failures should result in a rollback and not a commit." And someone needs to be able to say that with confidence before we can make this change, otherwise we could be breaking the evolutions code.

---

As already said, with transaction is intended to provide a managed connection, with commit on success and rollback on error. W/o such change I tend to decline this PR.

---

Hi, @stephendavidmarsh and @cchantep,

As a said [here](https://github.com/playframework/playframework/pull/8198#discussion_r165039470), a more conservative change here is to use `withConnection` which will handle the `setAutoCommit` possible errors, and we can also let evolutions code almost as is.

@stephendavidmarsh, I agree that there are a lot of possible improvements to be made at evolutions code. To be honest, we had already discussed the idea of using something like [Flyway](https://flywaydb.org/) in the future, but for now, the idea is to improve the API in a way that does not breaks user applications (or flow in development mode) at least for 2.6.x branch. That is why I've suggested [here](https://github.com/playframework/playframework/pull/8198#issuecomment-362002053) that we can use `withConnection` for 2.6.x and `withTransaction` for the master branch (where we have more room for changes).

Anyway, regarding "why did the unlock method silently swallow errors?" this apparently is because the code executed by lock/unlock is internal to Play and not part of application evolutions. As far as I remember, app evolutions will even be executed using another connection.

Regarding the extra `commit`, I'm failling to see how it is helpful here. If the block fails, `withTransaction` should rollback as expected. So, although it is a no-op as @stephendavidmarsh explained, it makes the code even more confusing.

> "Yes, the original code was wrong, failures should result in a rollback and not a commit."

You mean the evolutions code? Yes, failures should result in a rollback and not a commit is a good rule of thumb. How that justifies the additional commit?

---

Thank you, @marcospereira, for taking the time to respond clearly and at length.

> > "Yes, the original code was wrong, failures should result in a rollback and not a commit."

> You mean the evolutions code? Yes, failures should result in a rollback and not a commit is a good rule of thumb. How that justifies the additional commit?

It doesn't justify it, just the opposite in fact. If failures should result in a rollback, that additional commit does not need to be there. However, the original code would commit on failure. The additional commit was there to preserve the original behavior. It is definitely not a no-op.

I've removed the commits that made changes to the evolutions code from this PR. I think it would be best if improvements there waited until someone can give it the time and attention it needs.